File: solution_erg2/JAVA/Lab4.java

public class Lab4 {

    public static void main(String[] args) {

        int numThreads = 4; // local variable (in main) | not passed
	
	    int n = 0; // local variable (in main) |  pass-by-value | pass to instance variable that returns val to main (l. 18)
		CounterThread counterThreads[] = new CounterThread[numThreads]; // local variable (in main) | not passed
		for (int i = 0; i < numThreads; i++) { // local variable (in main) | pass-by-value
		    counterThreads[i] = new CounterThread(i, n); // pass-by-value arguments (i -> στοίχιση)
		    counterThreads[i].start();
		}
	
        for (int i = 0; i < numThreads; i++) {
            try {
				counterThreads[i].join();
				n += counterThreads[i].threadN;
            }
            catch (InterruptedException e) {}
		} 
		System.out.println("Main n = "+n); 
    }

}

class CounterThread extends Thread {
  	
    int threadID; // instance variable
    int threadN;  // instance variable
    
    public CounterThread(int tid, int n) { // pass-by-value parameters
        this.threadID = tid;
        this.threadN = n;
    }
  	
    public void run() {
              
        threadN = threadN + 1 + threadID;
        System.out.println("Thread "+threadID+ " n = "+threadN); 
    }
}

/* In this example, we don't have an array like in lab1,2 and 3, and instead we have
 * just a local variable n, which is passed by value to our thread class. In this case
 * however, the value is returned to main in line 18, and is added to the local variable
 * n, through the CounterThread instance variable threadN. I believe this case is the
 * same as Lab1, where since every instance has been initialized with 0, and especially
 * since n isn't global, the value inside every instance won't change. So we will have
 * an output like this:
 * 
 * T 3 n = 4
 * T 0 n = 1
 * T 1 n = 2
 * T 2 n = 3
 * Main n = 10
 * 
 * which proves to be the case. Example output: 
 * Thread 0 n = 1
 * Thread 2 n = 3
 * Thread 1 n = 2
 * Thread 3 n = 4
 * Main n = 10
 */

File: solution_erg2/C/Lab4.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

#define NTHREADS 4

typedef struct
{
	int threadID; // struct member
	int threadN;  // struct member
} thread_data;

void *counter_thread(void *ptr)
{
	int tID; // thread function var
	int *tN; // thread function var | pointer to threadN struct member

	thread_data *thread_args = (thread_data*) ptr;
	tID = thread_args->threadID; // value parameter
	tN  = &thread_args->threadN; // reference parameter

	*tN += 1 + tID;
	printf("Thread %d n = %d\n", tID, *tN);
}

int main()
{
	int i; // local main var | pass-by-value
	int n; // local main var | pass to instance variable that returns val to main
        pthread_t threads[NTHREADS]; // local main var | not passed
	thread_data thread_args[NTHREADS]; // local main var | pass-by-reference

	n = 0;
	for (i = 0; i < NTHREADS; i++)
	{
		thread_args[i].threadID = i;
		thread_args[i].threadN  = n;
		pthread_create(&threads[i], NULL, counter_thread, (void*) &thread_args[i]);
	}

	for (i = 0; i < NTHREADS; i++)
	{
		pthread_join(threads[i], NULL);
		n += thread_args[i].threadN;
	}

	printf("Main n = %d\n", n);

        return 0;
}
